Sprint 8

shepilovaElena opened the pull request:
добавление временных рамок задачи.
проверка на пересечение задач во времени(метод checkingIntersectionTwoTasks). метод getPrioritizedTasks.
тесты.

shepilovaElena committed:
тз-8.
добавление временных рамок задачи.
проверка на пересечение задач во времени(метод checkingIntersectionTwoTasks).
метод getPrioritizedTasks.
тесты.

shepilovaElena committed:
тз-8.
codestyle.

shepilovaElena committed:
тз-8.
codestyle.

shepilovaElena committed:
тз-8.
codestyle.

shepilovaElena committed:
тз-8.
codestyle.

irek-valiev commented on src/ru/yandex/practicum/javakanban/model/Task.java:
```diff
@@ -97,6 +136,10 @@ public void setId(int id) {
         this.id = id;
     }
 
+    public LocalDateTime getEndTime() {
```
добавить проверку на null для startTime, поскольку пользователь может создавать задачи не указывая сразу их время, он может пока их не знать

irek-valiev commented on src/ru/yandex/practicum/javakanban/manager/InMemoryTaskManager.java:
```diff
@@ -37,21 +44,94 @@ public Map<Integer, Subtask> getSubtasks() {
 
+    /// проверка перечечения временных интервалов задач
+    @Override
+    public Boolean checkingIntersectionTwoTasks(Task taskOne, Task taskTwo) {
```
метод не должен быть публичным и не должен быть в интерфейсе, так как является частью реализации и не должен быть виден или доступен для внешних потребителей

irek-valiev commented on src/ru/yandex/practicum/javakanban/manager/InMemoryTaskManager.java:
```diff
@@ -37,21 +44,94 @@ public Map<Integer, Subtask> getSubtasks() {
+        if (taskOne.getStartTime() != null || taskOne.getEndTime() != null || taskTwo.getStartTime() != null || taskTwo.getEndTime() != null) {
+            check = false;
+        }
+        if (taskTwo.getStartTime().isAfter(taskOne.getStartTime()) && taskTwo.getStartTime().isBefore(taskOne.getEndTime())
```
![Пересечение задач](https://github.com/user-attachments/assets/849dbef2-b231-49be-bc91-f4c55c255a24)

irek-valiev replied:
Приложил картинку всех возможных вариантов пересечения двух задач между собой. Проверь свое условие по ним и добавь недостающие

irek-valiev commented on src/ru/yandex/practicum/javakanban/manager/InMemoryTaskManager.java:
```diff
@@ -37,21 +44,94 @@ public Map<Integer, Subtask> getSubtasks() {
+        }
+        if (taskTwo.getStartTime().isAfter(taskOne.getStartTime()) && taskTwo.getStartTime().isBefore(taskOne.getEndTime())
+             || (taskOne.getStartTime().isAfter(taskTwo.getStartTime()) && taskOne.getStartTime().isBefore(taskTwo.getEndTime()))
+             || taskOne.getStartTime().equals(taskTwo.getStartTime())) {
```
это настолько невероятное событие, что его можно не ловить

irek-valiev commented on src/ru/yandex/practicum/javakanban/manager/InMemoryTaskManager.java:
```diff
@@ -37,21 +44,94 @@ public Map<Integer, Subtask> getSubtasks() {
+    }
+
+    @Override
+    public Boolean checkingIntersection(Task newTask) {
```
метод не должен быть публичным и не должен быть в интерфейсе, так как является частью реализации и не должен быть виден или доступен для внешних потребителей

irek-valiev commented on src/ru/yandex/practicum/javakanban/manager/InMemoryTaskManager.java:
```diff
@@ -37,21 +44,94 @@ public Map<Integer, Subtask> getSubtasks() {
+    @Override
+    public Boolean checkingIntersection(Task newTask) {
+       Optional<Task> intersection =  getPrioritizedTasks().stream()
+               .filter(task -> checkingIntersectionTwoTasks(task, newTask).equals(true))
```
.equals(true) лишнее, checkingIntersectionTwoTasks() итак возвращает boolean

irek-valiev commented on src/ru/yandex/practicum/javakanban/manager/InMemoryTaskManager.java:
```diff
@@ -37,21 +44,94 @@ public Map<Integer, Subtask> getSubtasks() {
+            endSubtask = (Subtask) subtasksArray[subtasksArray.length - 1];
+            epic.setEndTime(endSubtask.getEndTime());
+
+            epic.setDuration(Duration.between(firstSubtask.getStartTime(), endSubtask.getEndTime()));
```
продолжительность эпика - это сумма продолжительностей его подзадач

irek-valiev commented on src/ru/yandex/practicum/javakanban/manager/InMemoryTaskManager.java:
```diff
@@ -37,21 +44,94 @@ public Map<Integer, Subtask> getSubtasks() {
+
+        Subtask firstSubtask;
+        Subtask endSubtask;
+
```
не охвачен такой кейс, у эпика может быть множество подзадач, но ни у одной из них не заполнены атрибуты времени, т.е. subtasksArray.length будет 0. Тогда у эпика атрибуты времени должны быть null

irek-valiev commented on src/ru/yandex/practicum/javakanban/manager/InMemoryTaskManager.java:
```diff
@@ -63,18 +143,24 @@ public int addNewEpic(Epic newEpic) {
+                epic.addNewSubtask(newSubtask.getId());
+
+                subtasks.put(newSubtask.getId(), newSubtask);
+                changeEpicStatus(epics.get(newSubtask.getEpicId()));
```
сюда можно отдать epic, еще раз ходить за ним в хранилище не нужно

irek-valiev commented on src/ru/yandex/practicum/javakanban/manager/InMemoryTaskManager.java:
```diff
@@ -63,18 +143,24 @@ public int addNewEpic(Epic newEpic) {
+
+                subtasks.put(newSubtask.getId(), newSubtask);
+                changeEpicStatus(epics.get(newSubtask.getEpicId()));
+                changeEpicStartTimeEndTimeDuration(epics.get(newSubtask.getEpicId()));
```
аналогично

irek-valiev commented on src/ru/yandex/practicum/javakanban/manager/InMemoryTaskManager.java:
строчка 207 - нужно в prioritizedTasks удалить все задачи типа Task
строчка 219 - а здесь удалить в prioritizedTasks все задачи типа Subtask

irek-valiev commented on src/ru/yandex/practicum/javakanban/manager/InMemoryTaskManager.java:
```diff
@@ -150,12 +236,13 @@ public void deleteAllSubtasks() {
         for (Epic epic : epics.values()) {
```
не нужно разматывать epics.values() еще раз, все действия можно сделать за первый раз
+
и еще удалить в prioritizedTasks все задачи типа Subtask

irek-valiev commented on src/ru/yandex/practicum/javakanban/manager/InMemoryTaskManager.java:
```diff
@@ -227,17 +313,20 @@ public ArrayList<Subtask> getAllSubtasks() {
-                subtasks.remove(subtaskId);
-            }
+            epics.get(id).getEpicSubtasks().stream()
+                    .peek(i -> getPrioritizedTasks().remove(subtasks.get(i)))
```
не нужно использовать метод getPrioritizedTasks(), в нем собирается копия отсортированных задач. удаляй напрямую в prioritizedTasks

irek-valiev commented on src/ru/yandex/practicum/javakanban/manager/InMemoryTaskManager.java:
```diff
@@ -247,32 +336,36 @@ public void deleteEpicById(int id) {
     public void deleteSubtaskById(int id) {
         if (epics.containsKey(id)) {
             if (subtasks.containsKey(id)) {
+                getPrioritizedTasks().remove(subtasks.get(id));
```
аналогично

shepilovaElena committed:
тз-8.
исправление расчета duration для эпика.
исправление метода checkingIntersectionTwoTasks().
добавление в методы удаления задач и подзадач удаления их из prioritizedTasks.
